enterprise/provisionerd/remoteprovisioners.go

```diff
@@ -0,0 +1,461 @@
+	ctx      context.Context
+	executor Executor
+	cert     string
+	addr     string
+	listener net.Listener
+	logger   slog.Logger
+	tlsCfg   *tls.Config
+
+	mu      sync.Mutex
+	waiters map[string]waiter
+}
+
+func NewRemoteConnector(ctx context.Context, logger slog.Logger, exec Executor) (agpl.Connector, error) {
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, xerrors.Errorf("failed to listen: %w", err)
+	}
+	go func() {
+		<-ctx.Done()
+		ce := listener.Close()
+		logger.Debug(ctx, "listener closed", slog.Error(ce))
+	}()
+	r := &remoteConnector{
+		ctx:      ctx,
+		executor: exec,
+		listener: listener,
+		addr:     listener.Addr().String(),
+		logger:   logger,
+		waiters:  make(map[string]waiter),
+	}
+	err = r.genCert()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to generate certificate: %w", err)
+	}
+	go r.listenLoop()
+	return r, nil
+}
+
+func (r *remoteConnector) genCert() error {
+	privateKey, cert, err := GenCert()
+	if err != nil {
+		return err
+	}
+	r.cert = string(cert)
+	privateKeyBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		return xerrors.Errorf("failed to marshal private key: %w", err)
+	}
+	pkPEM := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privateKeyBytes})
+	certKey, err := tls.X509KeyPair(cert, pkPEM)
+	if err != nil {
+		return xerrors.Errorf("failed to create TLS certificate: %w", err)
+	}
+	r.tlsCfg = &tls.Config{Certificates: []tls.Certificate{certKey}, MinVersion: tls.VersionTLS13}
+	return nil
+}
+
+// GenCert is a helper function that generates a private key and certificate.  It
+// is exported so that we can test a certificate generated in exactly the same
+// way, but with a different private key.
+func GenCert() (*ecdsa.PrivateKey, []byte, error) {
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("generate private key: %w", err)
+	}
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: "Coder Provisioner Daemon",
+		},
+		DNSNames:  []string{serverName},
+		NotBefore: time.Now(),
+		// cert is valid for 5 years, which is much longer than we expect this
+		// process to stay up.  The idea is that the certificate is self-signed
+		// and is valid for as long as the daemon is up and starting new remote
+		// provisioners
+		NotAfter: time.Now().Add(time.Hour * 24 * 365 * 5),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("failed to create certificate: %w", err)
+	}
+	cert := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	return privateKey, cert, nil
+}
```

Comment:
nit: there's a few more instances where we generate a self-signed cert, we could perhaps extract a util function here.

Reply:
I'm not convinced it would save much trouble once we account for all the variable arguments